### PR TITLE
Community components

### LINKED ISSUE #1077: Enable community contributions for third-party components
I've heard of a few requests for third-party components that we would like to accept into Alloy, but are unable to provide support for based on the Grafana [release lifecycle](https://grafana.com/docs/release-life-cycle/). 

We should find a strategy for enabling the community to contribute these third-party components without it becoming subject to the Grafana release cycle.

### PR DESCRIPTION
**Pull request by wildum**
This PR introduces the concept of community components.

Community components are components owned and maintained by the community.
Grafana's commercial support offer does not cover them.

This category exists mainly to add vendor-specific components to Alloy like the Datadog exporter.

Closes #1077

### CODE REVIEW

File: docs/developer/adding-community-components.md
Comment (rfratto) [inline docs/developer/adding-community-components.md:5]: Open question: does "proposal" here mean the same thing as "proposal" in #909? 

If #909 gets accepted, I think we _should_ tie the two together, since proposals can be rejected for being technically infeasible, and that's one of the reasons we would not be able to accept a community component.

File: docs/developer/adding-community-components.md
Comment (rfratto) [inline docs/developer/adding-community-components.md:1]: Can we have a section briefly describing when something should be a community component vs an official component? 

I think mainly it's about vendors, right?

File: docs/developer/adding-community-components.md
Comment (rfratto) [inline docs/developer/adding-community-components.md:37]: I feel like the maintainer of community components should also be responsible for the documentation of that component. What do you think?

File: docs/sources/reference/cli/run.md
Comment (rfratto) [inline docs/sources/reference/cli/run.md:58]: I wonder if there's an existing convention for the flag names we could conform to here. For example, we have `--cluster.enabled` so `--community.enabled` or `--community-components.enabled` would align with that potentially.

File: docs/sources/concepts/community_components.md
Comment (rfratto) [inline docs/sources/concepts/community_components.md:18]: I think this is a bit strongly worded. My understanding is that we're worried about community components lacking an owner and preventing us from upgrading our dependencies or forcing us to act as the owners. 

Given that, I think this is the criteria for removal:

* The community component has been abandoned by its current owner, and 
* We are unable to find a new owner to take over, and 
* The unowned component is preventing us from being able to continue work on Alloy

I think that last bullet point is fairly important; it makes it clearer that we only remove things when it gets in the way of other progress.

File: docs/developer/adding-community-components.md
Comment (rfratto) [inline docs/developer/adding-community-components.md:18]: Should we talk about how "community components" is being rolled out slowly and we'll initially be slow to accept new components, so that we can refine the process and not get overwhelmed by dozens of new components?

**Review by rfratto — COMMENTED** [review]
I think this is turning out pretty well. I have a few questions about content first, but I'll have a second round of comments later to help wordsmith some things.

**Comment by wildum** [comment]
> Should this be labeled as a proposal?

This implementation follows a proposal that was approved through our previous review system. A new proposal isn't necessary unless we decide to revisit the concept or make fundamental changes.

**Comment by rfratto** [comment]
OK :+1: It might be easier to get consensus on the flag name via a proposal, but let's see how the next week goes for a natural discussion towards consensus

File: docs/sources/get-started/community_components.md
Comment (thampiotr) [inline docs/sources/get-started/community_components.md:1]: @clayton-cornell I thought it's worth checking with you: do you think this is the right place for this doc?

File: internal/alloycli/cmd_run.go
Comment (thampiotr) [inline internal/alloycli/cmd_run.go:168]:
```suggestion
	enableCommunityComps                    bool
```
I don't think the current name is clear enough and it will cause more cognitive overhead when reading the code.

File: internal/component/registry.go
Comment (thampiotr) [inline internal/component/registry.go:154]: Here the name `Community` is probably okay tho - there is enough context in the fact that this is inside the component Registration.

File: internal/component/registry.go
Comment (thampiotr) [inline internal/component/registry.go:178]:
```suggestion
		panic(fmt.Sprintf("Community component %q has a defined stability level - community components are not subject to this stability level setting. It should remain `undefined`", r.Name))
```

File: internal/runtime/alloy.go
Comment (thampiotr) [inline internal/runtime/alloy.go:111]: Not a fan of this name here either.

File: internal/runtime/internal/controller/component_registry.go
Comment (thampiotr) [inline internal/runtime/internal/controller/component_registry.go:24]: I don't think this is a good name. Does `community` mean that this registry is exclusively for community components? Does it mean that it's a community stability level? I think the naming should be more clear.

**Review by thampiotr — COMMENTED** [review]
Few comments about the naming etc. but no big blockers.

File: internal/component/registry.go
Comment (rfratto) [inline internal/component/registry.go:154]: silly nit: can we put this after Stability? It seems semantically related to stability.

File: docs/sources/get-started/community_components.md
Comment (rfratto) [inline docs/sources/get-started/community_components.md:1]: Some high level comments:

1. I think we should explicitly document on this page that community components do not have a stability level
2. I think we should modify the [backward compatibility page](https://grafana.com/docs/alloy/latest/introduction/backward-compatibility/) and write "community components" as an exception.

**Review by rfratto — COMMENTED** [review]
Basically LGTM, but I agree with Piotr that naming a boolean `Community` to indicate whether community components are enabled may not be named explicitly enough.

File: docs/developer/adding-community-components.md
Comment (wildum) [reply docs/developer/adding-community-components.md:35]: Should we just suggest to add the label proposal instead of using the review template?

File: .github/ISSUE_TEMPLATE/community_component_proposal.yaml
Comment (clayton-cornell) [inline .github/ISSUE_TEMPLATE/community_component_proposal.yaml:10]:
```suggestion
        Make sure you read the [Community Component Guidelines](https://github.com/grafana/alloy/tree/main/docs/developer/adding-community-components.md).
```

We thank them for opening the proposal and then tell them they need to do a thing BEFORE they open the proposal. We need to either remove the "Before..." part or find another way to phrase the "thank you..." part.

File: docs/developer/adding-community-components.md
Comment (clayton-cornell) [inline docs/developer/adding-community-components.md:7]: Are there other targets? or is it just for vendor-specific components?

File: docs/sources/get-started/community_components.md
Comment (clayton-cornell) [reply docs/sources/get-started/community_components.md:1]: I think it should be OK here. the `get-started` section is slowly being adapted and cleaned up and should explain what Alloy is in more detail... including a description of components in general. This new content fits in here since it just describes what they are... and doesn't provide any instructions or reference content.

File: docs/sources/get-started/community_components.md
Comment (clayton-cornell) [inline docs/sources/get-started/community_components.md:12]:
```suggestion
While Grafana does not offer commercial support for these components, they undergo acceptance and review by the {{< param "PRODUCT_NAME" >}} development team before being added to the repository.
```

We should not refer to Alloy in possessive (eg `Alloy's`)

File: docs/sources/get-started/community_components.md
Comment (clayton-cornell) [inline docs/sources/get-started/community_components.md:14]:
```suggestion
To use these community components, you must explicitly pass the `--feature.community-components.enabled` flag to the `run` command.
```

File: docs/sources/get-started/community_components.md
Comment (clayton-cornell) [inline docs/sources/get-started/community_components.md:17]:
```suggestion
__Community components__ without a maintainer may be disabled or removed if the components prevent or block the development of {{< param "PRODUCT_NAME" >}}.
```

I'm not sure about the wording here vs the intent.  A component without a maintainer could, in theory, block the release Alloy... if there's something that becomes incompatible with a new release?

File: docs/developer/adding-community-components.md
Comment (rfratto) [reply docs/developer/adding-community-components.md:35]: Users can't add labels themselves, but you can modify the issue template so that `proposal` is a default label when the issue template is used.

File: docs/developer/adding-community-components.md
Comment (wildum) [reply docs/developer/adding-community-components.md:7]: this is the primary target but we could also accept a vendor-neutral component as community component and not core component if we don't want to maintain it. I don't have an example for it and it's not something that we will do in the near future though (the first candidates for the community components are all vendor-specific)

File: docs/sources/get-started/community_components.md
Comment (wildum) [reply docs/sources/get-started/community_components.md:19]: that's correct. For example if by updating some dependencies we break a community component and there is no easy way to fix it + the community maintainers don't have time to do it then we might disable it/remove it to prevent from blocking a release or an important feature

File: docs/sources/get-started/community_components.md
Comment (clayton-cornell) [inline docs/sources/get-started/community_components.md:12]:
```suggestion
While Grafana does not offer commercial support for these components, they still undergo acceptance and review by the {{< param "PRODUCT_NAME" >}} development team before being added to the repository.
```

File: docs/sources/get-started/community_components.md
Comment (clayton-cornell) [inline docs/sources/get-started/community_components.md:16]:
```suggestion
__Community components__ don't have a stability level. They aren't covered by our [backward compatibility strategy][backward-compatibility].
```

**Review by clayton-cornell — APPROVED** [review]
A couple small tweaks for Vale linting.

**Review by rfratto — APPROVED** [review]
🎉

**Pull request merged**
